fgpyo/util/itertools.py

```diff
@@ -0,0 +1,195 @@
+    """
+    x: int
+    y: int
+
+    def incr_x(self) -> "Coord":
+        """Return a new Coord with the same y-coordinate and the x-coordinate incremented by 1."""
+        return Coord(self.x + 1, self.y)
+
+    def incr_y(self) -> "Coord":
+        """Return a new Coord with the same x-coordinate and the y-coordinate incremented by 1."""
+        return Coord(self.x, self.y + 1)
+
+
+@dataclass(order=True)
+class HeapElem(Generic[T]):
+    """
+    A heap element that contains a pairing of elements in the Cartesian product.
+
+    Attributes:
+        _priority: The priority of this element, which is the sum of the priorities of the paired
+            elements.
+        val: The paired elements.
+        coord: The position in the ProductSumTable.
+    """
+    _priority: float
+    val: Tuple[T, T] = field(compare=False)
+    coord: Coord  # NB: sorting must also be stable / deterministic on the coordinates
+
+    @classmethod
+    def new(cls, x: T, y: T, coord: Coord, priority: Callable[[T], float]) -> "HeapElem":
+        return cls(_priority=priority(x) + priority(y), val=(x, y), coord=coord)
+
+
+class ProductSumTable(Generic[T]):
+    """
+    A 2-D grid of elements that supports efficient traversal of the sorted Cartesian product.
+
+    Attributes:
+        xs: A sorted list of elements.
+        ys: A sorted list of elements.
+        priority: A function that returns a numeric priority for each element.
+    """
+    xs: List[T]
+    ys: List[T]
+    priority: Callable[[T], float]
+
+    def __init__(self, xs: Iterable[T], ys: Iterable[T], priority: Callable[[T], float]) -> None:
+        self.xs = list(xs)
+        self.ys = list(ys)
+        self.priority = priority
+
+    def has_right(self, elem: HeapElem) -> bool:
+        return (elem.coord.x + 1) < len(self.xs)
+
+    def has_down(self, elem: HeapElem) -> bool:
+        return (elem.coord.y + 1) < len(self.ys)
+
+    def is_last(self, elem: HeapElem) -> bool:
+        return (elem.coord.y + 1) == len(self.ys) and (elem.coord.x + 1) == len(self.xs)
+
+    def get_elem(self, coord: Coord) -> HeapElem:
+        x: T = self.xs[coord.x]
+        y: T = self.ys[coord.y]
+
+        return HeapElem.new(x=x, y=y, coord=coord, priority=self.priority)
+
+    def get_right(self, elem: HeapElem) -> HeapElem:
+        assert self.has_right(elem)
+        coord = elem.coord.incr_x()
+        return self.get_elem(coord)
+
+    def get_down(self, elem: HeapElem) -> HeapElem:
+        assert self.has_down(elem)
+        coord = elem.coord.incr_y()
+        return self.get_elem(coord)
+
+    def get_last(self) -> HeapElem:
+        x: T = self.xs[-1]
+        y: T = self.ys[-1]
+
+        return HeapElem.new(
+            x=x, y=y, coord=Coord(len(self.xs) - 1, len(self.ys) - 1), priority=self.priority
+        )
+
+
+def sorted_product(
+    xs: Collection[T],
+    ys: Collection[T],
+    priority: Callable[[T], float],
+    limit: Optional[int] = None,
+) -> List[Tuple[T, T]]:
```

Comment:
**FIXME** we should parametrize the algorithm used by the function, to enable the user to fall back to the naive approach (`sorted(itertools.product(xs, ys))`) for small inputs

also, someone sufficiently motivated could benchmark the two methods and identify a reasonable break even point